### ISSUE TITLE
Build Hearthstone server

### ISSUE THREAD

**Issue by Bert-Proesmans** [HearthSim/Firestarter#1]
Let's build an open-source HearthStone lobby server. 
**Notice** This issue and the accompanying discussion will close on Monday, 2018/07/16, at 21:00 CET.

The following points will act as cornerstone for this project.
- Support exactly one version of the client
The server will work with one specific version of the hearthstone client. This version will be selected based on amount of containing content and release date.
Currently investigating the use of version **6898 (v1.3)**, this is the latest patch of the Naxxramas cycle.

- Build a single-player experience
We will not facilitate PvP, playing against AI is the only option. 
This means starting off with adventures to be exact.
Currently thinking about only implementing the **Naxxramas adventure**.

- Increase interest in our Kettle protocol and game simulators
Polished bindings for related software improve the exposure and quality of said software.
Sabberstone, for example, lacks adventure card implementations. This server could be a reason for creating the missing implementations.

I encourage to discuss the mentioned points. Multiple opinions are necessary in order to refine their descriptions. Please keep answers on this issue to the point.
General talk also happens inside the stove channel on our discord, see https://hearthsim.info/join/.

**Edit**
A list of prominent features is shown below to clarify what the proposed target version will contain.

- Initial release features (basic, classic cardset, internationalization)
- Naxxramas adventure + card set 
- Arena mode
- Play mode
- Win 8+ support

What's not included.

- Tutorial ??
- Adventures, card-sets and card-backs from newer patches
- Any card changes (nerfs/buffs)
- Spectator mode (v2.0)
- Extended shop interface (v2.3)
- Alternate heroes (v2.7+)
- Tavern brawl mode (v2.7)
- [Game mode ID](https://hearthstone.gamepedia.com/Game_mode_ID) (v5.2)

**Edit**
The date for when the programming language will be decided is not yet set. We'll come back on this after this issue is closed.

**Comment by citiral** [comment]
I am in full support of this initiative.

Focusing on a single-player experience sounds like a good idea to me for several reasons. First of all, it will focus more development attention to AI development. Secondly, it emphasizes the difference between firestarter and the official lobby server, an argument we will hopefully never need, but you never know :)

I have some objection on the chosen version though. Focusing on a single adventure sounds like the right way to go, but I would suggest freezing the current version, and working from that. By developing on such an old version, you are needlessly constraining the possibilities you have, and if you ever want to expand to other adventures or gamemodes, a lot of work will have to be put in to upgrade the server. If we work on the currently latest version, we have a lot more space to be creative at little to no extra cost. Although if the encrypted libraries are a problem, it could be better to work on the latest unencrypted version (whichever that might be).

**Comment by Bert-Proesmans** [comment]
The version **6898 (v1.3)** has been picked for these exact reasons;
1. It doesn't require patching the game files itself (a HOSTS entry works fine)
2. It's old, which gives less reason (maybe?) for Blizzard to complain to us

Indirectly related;
- By being limited to a single adventure, we're able to "perfect" that functionality

*If these reasons sound wrong, please let me know*

I understand the benefit of supporting a later version is access to newer cards and adventures, but this point will circle back in a few years. Hearthstone is not a finished product and that newer version we choose will be "constraining the possibilities" in comparison to the future latest version.
The basics of the game remain the same: Play-, Arena-, Solo mode. We have room to repurpose these modes into something cool in any version.
I'm not saying that we should never think about supporting a newer version. Newer animations and content updates are nice to have, but it shouldn't be an objective.

In conclusion we should all agree on one specific version now and later when the added content becomes meaningful on its own.

About the mentioned version; I believe encryption was introduced with 19759 (v8.3), so the latest version without encryption would be 19632 (v8.2.2).
Check [this link](https://hearthstone.gamepedia.com/Patches) for more details about each patch.

**Comment by Bert-Proesmans** [comment]
This issue will close on Monday, 2018/07/16, at 21:00 CET. 
When this happens we will be going forward with the information mentioned in the first post. Until then discussion is still ongoing and the first post will be continuously updated.
Don't hesitate to discuss the cornerstones as described!

Also don't forget to open new issues suggesting a language to use for building the server, or leave your relevant comments there.

**Comment by Bert-Proesmans** [comment]
This issue is now closed and the proposed cornerstones are in effect.
Open a separate issue for further discussion on a specific part of these principles.

The deadline for proposing a programming language is one week from now, at Monday 2018/07/23 21:00 CET. Please see #2 for inspiration on how to create such an issue.
